### PR TITLE
JS: remove emptiness checks from the type confusion `x.length` sinks

### PR DESCRIPTION
**Pull request by esben-semmle**
This change eliminates the false positive of ODASA-7300.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/TypeConfusionThroughParameterTampering.qll [hunk 1]
```diff
@@ -101,7 +101,22 @@ module TypeConfusionThroughParameterTampering {
...
+          exists (ConditionGuardNode cond |
+            read.asExpr() = cond.getTest()
+          )
+          or
+          exists (EqualityTest eq, Expr zero |
+            zero.getIntValue() = 0 and
+            eq.hasOperands(read.asExpr(), zero)
+          )
+          or
+          exists (LogNotExpr eq |
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/security/dataflow/TypeConfusionThroughParameterTampering.qll]: `eq` is not a good variable name here

File: javascript/ql/src/semmle/javascript/security/dataflow/TypeConfusionThroughParameterTampering.qll [hunk 2]
```diff
@@ -101,7 +101,22 @@ module TypeConfusionThroughParameterTampering {
...
       exists (DataFlow::PropRead read |
-        read.accesses(this, "length")
+        read.accesses(this, "length") and
+        // exclude truthiness checks on the length: an array/string confusion can not control an emptiness check
+        not (
+          exists (ConditionGuardNode cond |
+            read.asExpr() = cond.getTest()
+          )
+          or
+          exists (EqualityTest eq, Expr zero |
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/security/dataflow/TypeConfusionThroughParameterTampering.qll]: I think any `Comparison` with zero should count as emptiness checks. Like, `if (xs.length > 0) { ...}`.

**Comment by esben-semmle** [comment]
Fixed

File: javascript/ql/src/semmle/javascript/security/dataflow/TypeConfusionThroughParameterTampering.qll [hunk 3]
```diff
@@ -101,7 +101,22 @@ module TypeConfusionThroughParameterTampering {
 
     LengthAccess() {
       exists (DataFlow::PropRead read |
-        read.accesses(this, "length")
+        read.accesses(this, "length") and
+        // exclude truthiness checks on the length: an array/string confusion can not control an emptiness check
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/security/dataflow/TypeConfusionThroughParameterTampering.qll]: s/can not/cannot/

**Comment by esben-semmle** [comment]
Fixed

**Pull request merged**